infer elementType for ranges in makeArray,makeSlice

wilzbach opened the pull request:
As discussed earlier, it's possible to infer the ElementType for `makeArray` with Ranges.
Even better we can keep full compatibility and make this feature opt-in - is anyone against this?

tl;dr: let's allow the variant b! (a is still possible)

```
auto a = alloc.makeArray!long([5, 42]);
auto b = alloc.makeArray([5, 42]);
```

Ping @9il @andralex

9il commented on std/experimental/allocator/package.d:
```diff
@@ -748,31 +748,40 @@ unittest
 
 /// Ditto
-T[] makeArray(T, Allocator, R)(auto ref Allocator alloc, R range)
+auto makeArray(T = void, Allocator, R)(auto ref Allocator alloc, R range)
```
Why not `T = ElementType!Range`?

wilzbach replied:
I would prefer that, but
1) this might break compatibility (which might be okay because it's experimental)
2) removes to specify the feature to use a different type

9il replied:
`auto makeArray(T = ElementType!R, Allocator, R)`
1 and 2. really?

wilzbach commented:
thanks to @9il helpful I removed `T = void` and replaced it with an additional overload without `T`.

9il commented on std/experimental/allocator/package.d:
```diff
@@ -748,6 +748,15 @@ unittest
+Unqual!(ElementType!R)[] makeArray(Allocator, R)(auto ref Allocator alloc, R range)
+if (isInputRange!R)
+{
+    import std.range;
```
`.primitives`

9il commented on std/experimental/ndslice/slice.d:
```diff
@@ -681,6 +691,19 @@ auto makeSlice(T,
+    import std.experimental.allocator;
+    import std.experimental.allocator.mallocator;
+
+    // also works with 'traditional' makeArray API
```
both api would be _traditional_

wilzbach replied:
changed to `double` to show-off casting.

ZombineDev commented:
I also prefer Ilya's version:
`auto makeArray(T = ElementType!R, Allocator, R)`
1. Doesn't break the API.
2. Still allows the user to specify the type manually.

wilzbach commented:
> auto makeArray(T = ElementType!R, Allocator, R)

I would prefer that too, but unfortunately D isn't that smart. It yields `R is not defined`.

MetaLang commented:
@wilzbach that *should* work. It may be a possible bug with template argument type deduction.

wilzbach commented:
> @wilzbach that should work. It may be a possible bug with template argument type deduction.

@MetaLang AFAIK it's a missing feature, which would be awesome to have!
See [Issue 12433](https://issues.dlang.org/show_bug.cgi?id=12433)

andralex commented on std/experimental/allocator/package.d:
```diff
@@ -748,6 +748,15 @@ unittest
+if (isInputRange!R)
+{
+    import std.range.primitives;
+    alias T = Unqual!(ElementType!R);
```
Use `ElementEncodingType`?

wilzbach replied:
[ElementEncodingType](https://github.com/dlang/phobos/blob/v2.071.0/std/range/primitives.d#L1183):

```d
    static if (is(StringTypeOf!R) && is(R : E[], E))
        alias ElementEncodingType = E;
    else
        alias ElementEncodingType = ElementType!R
```

[ElementType](https://github.com/dlang/phobos/blob/v2.071.0/std/range/primitives.d#L1097):

```d
    static if (is(typeof(R.init.front.init) T))
        alias ElementType = T;
    else
        alias ElementType = void;
```

... where is the difference?

ZombineDev replied:
ElementType!(char[]) == dchar
ElementEncodingType!(char[]) == char

andralex commented on std/experimental/ndslice/slice.d:
```diff
@@ -638,6 +638,16 @@ auto makeSlice(T,
 
 /// ditto
+auto makeSlice(
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
```
So... where can I find docs for replaceArrayWithPointer? https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=replacearraywithpointer doesn't help.

wilzbach replied:
It's dittoed from the method above, which is part of the upcoming ndslice modifications for 2.072. [Preview here](http://dlang.org/phobos-prerelease/std_experimental_ndslice_slice.html)

@9il should we explain these flags in the slice module header?

9il replied:
yes

9il replied:
@wilzbach please add `replaceArrayWithPointer` for each function in the module and fix its name in the docs for `sliced` (it called `mod` instead of `replaceArrayWithPointer`)

andralex commented:
preapproved once nits are looked at

wilzbach commented:
> (Because R.init.front.init is dchar for narrow strings.)

Ah auto-decoding :S 
Thanks for the explanation @ZombineDev!
(Rebased)

ZombineDev commented:
No problem :)

Maybe not directly related to your PR, but do you happen to know what will `makeArray` return if the elements of the range are classes - an array of references to class instances or the actual array where the objects were allocated and emplaced?

JackStouffer commented:
Ping @9il, I think this can be merged now

9il commented on std/experimental/allocator/package.d:
```diff
@@ -734,6 +734,15 @@ unittest
 
 /// Ditto
+Unqual!(ElementType!R)[] makeArray(Allocator, R)(auto ref Allocator alloc, R range)
+if (isInputRange!R)
```
`&& !infinityRange!R`

9il commented on std/experimental/allocator/package.d:
```diff
@@ -734,6 +734,15 @@ unittest
 }
 
 /// Ditto
+Unqual!(ElementType!R)[] makeArray(Allocator, R)(auto ref Allocator alloc, R range)
```
Add unittest for strings please

wilzbach commented on std/experimental/allocator/package.d:
```diff
@@ -763,10 +772,18 @@ if (isInputRange!R && !isInfinite!R)
 
-        for (; !range.empty; range.popFront, ++i)
+        import std.conv : emplace;
+        static if (isNarrowString!R || isRandomAccessRange!R)
```
we need the special case for narrow strings & as we already have the special case, using it for arrays will be faster too (see e.g. #4265)

wilzbach commented:
@9il addressed your comments - you were correct: strings are evil :S 
I hope the special-casing I inserted is correct.

> @wilzbach please add replaceArrayWithPointer for each function in the module and fix its name in the docs for sliced 

Will do this in a separate PR ;-)

9il commented:
LGTM

9il commented:
Travis fails, fix/rebase please

schveiguy commented:
@9il FYI, you can log in your GH user to travis to force a rebuild (as a dlang member). Just figured this out recently. However, in this case, I don't think that will help, as there is an actual issue.

wilzbach commented:
> Travis fails, fix/rebase please

Done. Should be passing now.

9il commented:
Auto-merge toggled on
